src/jit/smallhash.h

```diff
@@ -0,0 +1,605 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef _SMALLHASHTABLE_H_
+#define _SMALLHASHTABLE_H_
+
+//------------------------------------------------------------------------
+// HashTableInfo: a concept that provides equality and hashing methods for
+//                a particular key type. Used by HashTableBase and its
+//                subclasses.
+template<typename TKey>
+struct HashTableInfo
+{
+    // static bool Equals(const TKey& x, const TKey& y);
+    // static unsigned GetHashCode(const TKey& key);
+};
+
+//------------------------------------------------------------------------
+// HashTableInfo<TKey*>: specialized version of HashTableInfo for pointer-
+//                       typed keys.
+template<typename TKey>
+struct HashTableInfo<TKey*>
+{
+private:
+    // Recursive template definition to calculate the base-2 logarithm
+    // of a constant value.
+    template <unsigned val, unsigned acc = 0>
+    struct Log2 { enum { value = Log2<val / 2, acc + 1>::value }; };
+
+    template <unsigned acc>
+    struct Log2<0, acc> { enum { value = acc }; };
+
+    template <unsigned acc>
+    struct Log2<1, acc> { enum { value = acc }; };
+
+public:
+    static bool Equals(const TKey* x, const TKey* y)
+    {
+        return x == y;
+    }
+
+    static unsigned GetHashCode(const TKey* key)
+    {
+        // Shift off bits that are not likely to be significant
+        size_t keyval = reinterpret_cast<size_t>(key) >> Log2<__alignof(TKey)>::value;
```

Comment:
> Given how cheap hashing is for pointers perhaps we can improve the memory usage a bit by not storing the hash in the bucket.

This is an interesting idea. In that case I can even imagine only using the hash to find the home bucket and then simply comparing against the values at each element in the chain. This idea would then only require hash recomputation in `Resize()`.

> Given how cheap hashing is for pointers perhaps we can improve the memory usage a bit by not storing the hash in the bucket. And if we don't need nullptr keys then we can treat a bucket with a nullptr key as empty and get rid of m_isFull.

Another option is to shorten `m_firstOffset` and `m_isFull` to 31 and 1 bits in width, respectively.